### PR TITLE
Shifted existing implementations to Optax and implemented Meta SGD

### LINKED ISSUE #15: Implementation: Meta-SGD
https://arxiv.org/abs/1707.09835

### LINKED ISSUE #17: shift to optax optimizers
Flax optimizers have been [deprecated](https://flax.readthedocs.io/en/latest/flax.optim.html) in favour of optax optimizers. We need to make the shift in the existing implementation(s)

### PR DESCRIPTION
**Pull request by AnirudhM1**
Fixes #17
Fixes #15 

- Optax optimizers are used instead of flax.optim. 
- Added Meta SGD. 

---
Example Code Usage:
```python
import flax.linen as nn
class MLP(nn.Module):
    @nn.compact:
    def __call__(self, x, train=True):
        return nn.Dense(10)(x)
```
```python
from opti_trainer import OptiTrainer
from maml import maml_adapt

params = mlp.init(init_key, init_arr)
tx = optax.sgd(0.01) # Outer optimizer
state = OptiTrainer.create(params=params, apply_fn=mlp.apply, adapt_fn=maml_adapt, loss_fn=mse, tx=tx)

for epoch in range(100):
    total_loss = []
    for tasks in task_loader:
        state, loss, metrics_arr = OptiTrainer.meta_train_step(state, tasks, metrics=(accuracy, f1_score))
        total_loss.append(loss)
    loss = jnp.mean(jnp.stack(total_loss))
    print(f'Epoch {epoch+1}/{100}: {loss}')
```

### CODE REVIEW

File: jeta/opti_trainer.py [hunk 1]
```diff
@@ -1,76 +1,111 @@
...
-        Returns
-        -------
-        Mean of task losses.
+        Args:
+            params (flax.core.FrozenDict[str, Any]): Parameters of the model.
+            apply_fn ((params, x) -> y): Function that applies the model to a batch of data.
+            adapt_fn ((params, apply_fn, loss_fn, support_set) -> adapted_params): Specific meta learning function
+            which adapts to the support set.
+            loss_fn ((logits, targets) -> loss): Loss Function.
+            tx (Optax Optimizer): Optax optimizer.
```
Comment (threewisemonkeys-as) [inline jeta/opti_trainer.py]:
```suggestion
            tasks ((x_train, y_train), (x_test, y_test)): Batch of tasks to be trained on
```

File: jeta/opti_trainer.py [hunk 2]
```diff
@@ -1,76 +1,111 @@
...
-    def loss_fn(model):
+    def meta_test_step(state: MetaTrainState, tasks) -> ndarray:
+        """Performs a single meta-testing step on a batch of tasks.
+
+        The function first adapts to the support set and then evaluates it's perfomance
+        on the query set.
+
+        Args:
+            state (MetaTrainState): Contains information regarding the current state.
+            tasks ((x_train, y_train), (x_test, y_test)): Batch of tasks to be trained on.
```
Comment (threewisemonkeys-as) [inline jeta/opti_trainer.py]:
```suggestion
            tasks ((x_train, y_train), (x_test, y_test)): Batch of tasks to be evaluated on.
```

**Review by threewisemonkeys-as — CHANGES_REQUESTED** [review]
Some minor comments, otherwise looks good to me. 

@AnirudhM1 have you tried this out on any dataset?

**Comment by AnirudhM1** [comment]
@threewisemonkeys-as Yes. I have tested this on the _Sin dataset_  as a regression task as mentioned in the [Meta SGD paper](https://arxiv.org/abs/1707.09835) and have obtained results similar to the ones reported in the paper for both MAML and Meta SGD.

**Comment by veds12** [comment]
@AnirudhM1 let's wait till #21 gets merged. Should take a couple of days. Till then, feel free to take a look at other issues :)

**Comment by veds12** [comment]
@AnirudhM1 PR #21 has been merged. Can you run the pre-commit and make sure all the tests pass

**Comment by AnirudhM1** [comment]
> ```python
> def __call__(self, x, train=True):
>     return nn.Dense(10)(x)
> ```

---

Here using the kwarg `train` is necessary regardless of it being used or not. This is because layers like `batchnorm` behave differently at train/test time and need some form of indicator. 
To make the interface consistent across all types of models, kwarg 'train' is assumed to be present in all models.

 

To enable consistent use of models, some minor changes have been made.
  
cc: @veds12 @threewisemonkeys-as

**Comment by veds12** [comment]
@AnirudhM1 Lint is still failing. Are you running pre-commit before pushing?
